doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1746,33 +1784,77 @@ Controlling the proof flow
    impossible. In most cases, :g:`P` or :g:`∼P` is one of the hypotheses of the
    local context.
 
-.. tacn:: contradiction
-   :name: contradiction
+.. tacn:: contradiction {? @one_term_with_bindings }
 
-   This tactic applies to any goal. The contradiction tactic attempts to
-   find in the current context (after all intros) a hypothesis that is
-   equivalent to an empty inductive type (e.g. :g:`False`), to the negation of
-   a singleton inductive type (e.g. :g:`True` or :g:`x=x`), or two contradictory
-   hypotheses :g:`P` and :g:`not P`.
+   Tries to prove the current goal by finding a contradiction.
 
-   .. exn:: No such assumption.
-      :undocumented:
+   If :n:`@one_term_with_bindings` is not provided, the tactic first does an
+   :tacn:`intros`.  The tactic then proves the goal if the updated context has
+   a pair of hypotheses where one is the negation of the other (e.g. :n:`P` and
+   not :n:`~P`) or if there is a hypothesis with the type :n:`False` (or another
+   empty inductive type).
+
+   If :n:`@one_term_with_bindings` is provided: The type of
+   :n:`@one_term_with_bindings` must be a negation, such as :n:`~P`,
+   or :n:`False` (or another empty inductive type).
+   If the type is a negation and :n:`P` is a hypothesis in the context,
+   the goal is proven.  If the type is a negation and :n:`P` is not in
+   the context, the goal is replaced with :n:`P`.  If the type is :n:`False`,
+   the goal is proven.  Otherwise the tactic fails.  (In particular, note
+   that the converse doesn't work, e.g. if the type is :n:`P`, the tactic
+   won't find :n:`~P`.)
+
+   Use the :tacn:`discriminate` tactic to prove the current goal when there
+   is a hypothesis with an impossible structural equality such as
+   :n:`0 = 1`.
 
-.. tacv:: contradiction @ident
+.. example:: :tacn:`contradiction` tactic
+
+   Simple examples.  To see more detail, add `intros` after each `Goal`.
+
+   .. coqtop:: reset in
+
+      Inductive F :=. (* Another empty inductive type *)
+
+      Goal F -> False.
+      contradiction.
+      Qed.
+
+      Goal forall (A : Prop), A -> ~A -> False.
+      contradiction.
+      Qed.
+
+      Goal forall (A : Type) (x : A), ~(x = x) -> False.
+      contradiction.
+      Qed.
+
+   Apply a fact from the standard library:
+
+   .. coqtop:: in
+
+      Require Import Arith.
+
+      Goal forall (A : Prop), 0 < 0 -> A.
+
+   .. coqtop:: all
 
-   The proof of False is searched in the hypothesis named :n:`@ident`.
+      intros. (* the intros is needed because 0 < 0 is not a negation
+                 while the type of (Nat.lt_ irrefl 0) is a negation (~ 0 < 0) *)
```

Comment:
The explanation from the comment is not the good one. It's just that `contradiction` never does an `intros` when used with an argument.